plugins/GSdx/GSRendererDX.cpp

```diff
@@ -48,15 +48,139 @@ GSRendererDX::~GSRendererDX()
+		ps_cb.FogColor_AREF.a = (float)context->TEST.AREF - 0.1f + 1.0f;
+		ps_sel.atst = 2;
+		break;
+	case ATST_EQUAL:
+		ps_cb.FogColor_AREF.a = (float)context->TEST.AREF;
+		ps_sel.atst = 3;
+		break;
+	case ATST_NOTEQUAL:
+		ps_cb.FogColor_AREF.a = (float)context->TEST.AREF;
+		ps_sel.atst = 4;
+		break;
+	case ATST_NEVER:
+	case ATST_ALWAYS:
+	default:
+		ps_sel.atst = 0;
+		break;
+	}
+
+	// Destination alpha pseudo stencil hack: use a stencil operation combined with an alpha test
+	// to only draw pixels which would cause the destination alpha test to fail in the future once.
+	// Unfortunately this also means only drawing those pixels at all, which is why this is a hack.
+	// The interaction with FBA in D3D9 is probably less than ideal.
+	if (UserHacks_AlphaStencil && DATE && dev->HasStencil() && om_bsel.wa && (!context->TEST.ATE || context->TEST.ATST == 1))
+	{
+		if (!context->FBA.FBA)
+		{
+			if (context->TEST.DATM == 0)
+				ps_sel.atst = 2; // >=
+			else {
+				if (tex && tex->m_spritehack_t)
+					ps_sel.atst = 0; // <
+				else
+					ps_sel.atst = 1; // <
+			}
+			ps_cb.FogColor_AREF.a = (float)0x80;
+		}
+		if (!(context->FBA.FBA && context->TEST.DATM == 1))
+			om_dssel.alpha_stencil = 1;
+	}
+}
+
+void GSRendererDX::EmulateZbuffer()
+{
+	if (context->TEST.ZTE)
+	{
+		om_dssel.ztst = context->TEST.ZTST;
+		om_dssel.zwe = !context->ZBUF.ZMSK;
+	}
+	else
+	{
+		om_dssel.ztst = ZTST_ALWAYS;
+	}
+
+	uint32 max_z;
+	if (context->ZBUF.PSM == PSM_PSMZ32) {
+		max_z = 0xFFFFFFFF;
+	}
+	else if (context->ZBUF.PSM == PSM_PSMZ24) {
+		max_z = 0xFFFFFF;
+	}
+	else {
+		max_z = 0xFFFF;
+	}
+
+	// The real GS appears to do no masking based on the Z buffer format and writing larger Z values
+	// than the buffer supports seems to be an error condition on the real GS, causing it to crash.
+	// We are probably receiving bad coordinates from VU1 in these cases.
+
+	if (om_dssel.ztst >= ZTST_ALWAYS && om_dssel.zwe && (context->ZBUF.PSM != PSM_PSMZ32)) {
+		if (m_vt.m_max.p.z > max_z) {
+			ASSERT(m_vt.m_min.p.z > max_z); // sfex capcom logo
+			// Fixme :Following conditional fixes some dialog frame in Wild Arms 3, but may not be what was intended.
+			if (m_vt.m_min.p.z > max_z) {
+#ifdef _DEBUG
+				fprintf("Bad Z size on %s buffers", psm_str(context->ZBUF.PSM));
+#endif
+				om_dssel.ztst = ZTST_ALWAYS;
+			}
+		}
+	}
+
+	GSVertex* v = &m_vertex.buff[0];
+	// Minor optimization of a corner case (it allow to better emulate some alpha test effects)
+	if (om_dssel.ztst == ZTST_GEQUAL && (m_vt.m_eq.xyzf & 0x8) && v[0].XYZ.Z == max_z) {
+#ifdef _DEBUG
+		fprintf("Optimize Z test GEQUAL to ALWAYS (%s)", psm_str(context->ZBUF.PSM));
+#endif
+		om_dssel.ztst = ZTST_ALWAYS;
+	}
+}
+
 void GSRendererDX::DrawPrims(GSTexture* rt, GSTexture* ds, GSTextureCache::Source* tex)
 {
-	GSDrawingEnvironment& env = m_env;
-	GSDrawingContext* context = m_context;
+	context = m_context;
+	env = m_env;
```

Comment:
I mean the only pupose of `env` is to write `env.foo` instead of `m_env.foo`. In short it is useless. If you use `env...` in the new created function for depth/alpha, you just need to call `m_env....` instead.